File: Niveau 4/Graphes implicites 1/Arbres malades/main.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <utility>
#include <tuple>
#include <numeric>
int dfs(const std::vector<std::vector<int>> &graph, int i)
{
  int n = graph.size();
  std::vector<int> visited(n, 0);
  visited [i] = 1;
  std::stack<int> to_visit;
  to_visit.push(i);
  while (!to_visit.empty())
  {
    int u = to_visit.top();
    to_visit.pop();
    for (int v : graph [u])
    {
      if (!visited [v])
      {
        visited [v] = 1;
        to_visit.push(v);
      }
    }
  }
  return std::accumulate(visited.begin(), visited.end(), 0);
}
std::vector<std::vector<int>> buildGraph(const std::vector<std::tuple<int, int, int>> &arbres)
{
  int n = arbres.size();
  std::vector<std::vector<int>> graph(n);
  for (int i {0}; i < n; ++i)
  {
    int xA, yA, rA;
    std::tie(xA, yA, rA) = arbres [i];
    for (int j {0}; j < n; ++j)
    {
      if (i != j)
      {
        int xB, yB, rB;
        std::tie(xB, yB, rB) = arbres [j];
        int dd = (xA - xB) * (xA - xB) + (yA - yB) * (yA - yB);
        if (dd <= rA * rA)
        {
          graph [i].push_back(j);
        }
      }
    }
  }
  return graph;
}
std::pair<std::vector<std::tuple<int, int, int>>, std::vector<int>> capture()
{
  int N, M; // N = nombre d'arbres, M = nombre de questions
  std::cin >> N >> M;
  std::vector<std::tuple<int, int, int>> arbres(N);
  for (int i {0}; i < N; ++i)
  {
    int x, y, r;
    std::cin >> x >> y >> r;
    arbres [i] = std::make_tuple(x, y, r);
  }
  std::vector<int> questions(M);
  for (int i {0}; i < M; ++i)
  {
    std::cin >> questions [i];
  }
  return std::make_pair(arbres, questions);
}
int main()
{
  std::ios::sync_with_stdio(false);
  auto data = capture();
  std::vector<std::tuple<int, int, int>> arbres = data.first;
  std::vector<int> questions = data.second;
  int n = arbres.size();
  std::vector<std::vector<int>> graph = buildGraph(arbres);
  std::vector<int> reponses(n);
  for (int i {0}; i < n; ++i)
  {
    reponses [i] = dfs(graph, i);
  }
  for (int i : questions)
  {
    std::cout << reponses [i] << '\n';
  }
  return 0;
}

File: Niveau 4/Graphes implicites 1/Arbres malades/main.py
import sys
# Construction du graphe basé sur les arbres et leur rayon de contamination
def build_graph(arbres):
    n = len(arbres)
    graph = [[] for _ in range(n)]
    bidirect = [[] for _ in range(n)]
    # Remplissage du graphe en vérifiant la bidirectionnalité de la contamination
    for i in range(n):
        xA, yA, rA = arbres[i]
        for j in range(i + 1, n):
            xB, yB, rB = arbres[j]
            dist2 = (xA - xB) * (xA - xB) + (yA - yB) * (yA - yB)
            arbre2VersArbre1 = dist2 <= rB * rB
            if dist2 <= rA * rA:  # A peut contaminer B
                graph[i].append(j)
                if arbre2VersArbre1:
                    bidirect[i].append(j)
            if arbre2VersArbre1:  # B peut contaminer A
                graph[j].append(i)
    return graph, bidirect
# Fonction DFS pour propager la contamination à partir d'un arbre donné
def dfs(graph, start):
    n = len(graph)
    visited = [False] * n
    visited[start] = True
    stack = [start]  # Utiliser une pile pour le DFS
    count = 1  # Compter l'arbre lui-même
    while stack:
        u = stack.pop()  # Récupérer le dernier élément ajouté (LIFO)
        for v in graph[u]:
            if not visited[v]:
                visited[v] = True
                stack.append(v)
                count += 1
    return count
def main():
    # Capture des données
    N, M = map(
        int, sys.stdin.readline().split()
    )  # N = nombre d'arbres, M = nombre de questions
    # Lire les arbres
    arbres = []
    for _ in range(N):
        x, y, r = map(int, sys.stdin.readline().split())
        arbres.append((x, y, r))
    # Construction du graphe
    graph, bidirect = build_graph(arbres)
    reponses = [-1] * N  # Initialisation à -1 pour marquer les non-calculés
    # Traiter chaque question
    for _ in range(M):
        q = int(sys.stdin.readline())
        if reponses[q] != -1:
            sys.stdout.write(
                str(reponses[q]) + "\n"
            )  # Imprimer immédiatement si déjà calculé
        else:
            # Calculer la réponse uniquement pour cet arbre avec DFS
            reponse = dfs(graph, q)
            reponses[q] = reponse  # Mémoriser uniquement pour cet arbre
            for a in bidirect[q]:
                reponses[a] = reponse
            sys.stdout.write(str(reponse) + "\n")  # Imprimer immédiatement la réponse
if __name__ == "__main__":
    main()
